backend/pkg/rate/limit_http_test.go

```diff
@@ -0,0 +1,181 @@
+			t.Errorf("expected status code %d, received: %d", http.StatusOK, w.Code)
+		}
+	})
+
+	t.Run("Reserve", func(t *testing.T) {
+		t.Parallel()
+		reqFast, _ := http.NewRequest("GET", "http://localhost/fast/apis", nil)
+		res, _ := limiter.Reserve(reqFast)
+		if !res.OK() {
+			t.Error("unexpected rate limit on first request to fast rate limiter")
+		} else {
+			time.Sleep(time.Microsecond)
+			res, _ = limiter.Reserve(reqFast)
+			if !res.OK() {
+				t.Error("unexpected rate limit after waiting on fast request")
+			}
+		}
+
+		reqSuperSlow, _ := http.NewRequest("POST", "http://localhost/slow/superslow", nil)
+		res, _ = limiter.Reserve(reqSuperSlow)
+		if !res.OK() {
+			t.Errorf("unexpected rate limit for superslow request")
+		} else {
+			res, _ = limiter.Reserve(reqSuperSlow)
+			if res.OK() {
+				t.Errorf("request to superslow should get rate limited")
+			} else if res.Delay() < time.Hour*23 && res.Delay() > time.Hour*25 {
+				t.Errorf("superslow request delay is not close to 24h: actual: %s", res.Delay())
+			}
+		}
+		reqOkieDokie, _ := http.NewRequest("GET", "http://localhost/okiedokie/foobar", nil)
+		res, _ = limiter.Reserve(reqOkieDokie)
+		if _, ok := res.(okReservation); !ok {
+			t.Errorf("expected *okReservation, received: %T", res)
+		} else if !res.OK() {
+			t.Error("the request was not expected to be rate limited")
+		}
+	})
+	const jwt = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9." +
+		"eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9" +
+		"lIiwiYWRtaW4iOnRydWUsImlhdCI6MTUxNjIzOTAyMn0." +
+		"KMUFsIDTnFmyG3nMiGM6H9FNFUROf3wh7SmqJp-QV30"
+	t.Run("MiddlewareGin", func(t *testing.T) {
+		limiterGin := newHTTPLimiterForTesting(t)
+		handler := gin.New()
+		handler.Use(accesslog.Middleware(), identity.Middleware(), limiterGin.MiddlewareGin)
+		handler.Handle("GET", "/subject/foo", func(ctx *gin.Context) { ctx.Status(http.StatusOK) })
+		req, _ := http.NewRequest("GET", "http://localhost/subject/foo", nil)
+		req.Header.Set("Authorization", "Bearer "+jwt)
+		for range 5 {
+			w := httptest.NewRecorder()
+			handler.ServeHTTP(w, req)
+			if w.Code != http.StatusOK {
+				t.Errorf("unexpected status code: %d, expected: %d", w.Code, http.StatusOK)
+			}
+		}
+		w := httptest.NewRecorder()
+		handler.ServeHTTP(w, req)
+		if w.Code != http.StatusTooManyRequests {
+			t.Errorf("unexpected status code: %d, expected: %d", w.Code, http.StatusTooManyRequests)
+		}
+		req.URL.Path = "/bad_event/"
+		w = httptest.NewRecorder()
+		handler.ServeHTTP(w, req)
+		if w.Code != http.StatusInternalServerError {
+			t.Errorf("unexpected status code: %d, expected: %d", w.Code, http.StatusInternalServerError)
+			t.Error("expected template execution to fail")
+		}
+		req.URL.Path = "/bad_group/"
+		w = httptest.NewRecorder()
+		handler.ServeHTTP(w, req)
+		if w.Code != http.StatusInternalServerError {
+			t.Errorf("unexpected status code: %d, expected: %d", w.Code, http.StatusInternalServerError)
+			t.Error("expected template execution to fail")
```

Comment:
the execution will never reach this line, Errorf calls Fail() or at least thats what I think.

Reply:
ah, yes, I think the fact that Error[f]* calls Fail() vs Fatal calling FailNow helped me here. sorry for the noise.